R/estimateDominance.R

```diff
@@ -0,0 +1,366 @@
+
+              #If the indices list is not null
+              if (!is.null(index)) {
+                  #loop through index list
+                  for (idx in index) {
+                      #Get the specific index, and save it to table
+                      tab <- cbind(tab,
+                          dominance_help(x,
+                              abund_values=abund_values, index=idx, rank=rank,
+                              as_relative=as_relative,
+                              aggregate=aggregate))
+                  }
+          
+                  # Add indices names to table's columns' names
+                  colnames(tab) <- index
+          
+                  # Convert table to data frame
+                  tab <- as.data.frame(tab)
+
+                  # Adds index data to original MicrobiomeExperiment object
+                  if (length(name) == 1 && name == "all") {
+                      name <- colnames(tab)
+                  } 
+                  x <- add_indices_to_coldata(x, tab, name)
+              }
+
+              # Return ME object
+              return(x)
+
+
+          }
+)
+
+
+
+
+#---------------------------Help functions----------------------------------------------------------------
+checkIndices <- function(index){
+
+    # Only include accepted indices
+    #If index is not null, save them to "index" variable after changing them to lower case
+    if (!is.null(index)) {index <- tolower(index)}
+
+    #Saves the list of accepted indices to "accepted" variable
+    accepted <- tolower(c("DBP", "DMN", "absolute", "relative",
+                          "simpson", "core_abundance", "gini"))
+
+    # Return all indices
+    #If the user has written in the "index" argument "all"
+    if (length(index) == 1 && index == "all") {
+        index <- accepted
+    }
+
+    #If there is indices, but none of them are accepted, gives warning and returns null
+    if (!is.null(index) && !any(index %in% accepted)) {
+        warning(paste("None of the dominance indices are not recognized:", paste(index, sep = ",")))
+        return(NULL)
+    }
+
+    #If there are indices, take only those indices that are accepted, and save them to variable "index"
+    if (!is.null(index)) {
+
+        #Stores indices that are not accepted
+        discarded <- setdiff(index, accepted)
+
+        #If there are indcices that are not accepted, gives a warning.
+        if(!rlang::is_empty(discarded)){
+            warning(paste("The following dominance indices are not recognized:", paste(discarded, sep = ",")))
+        }
+        #Stores indices that are accepted
+        index <- intersect(index, accepted)
+    }
+
+    return(index)
+}
+
+dominance_help <- function(x, abund_values = "counts", index="all", rank=1, as_relative=TRUE,
+                           aggregate=TRUE) {
+    #Stores the absolute abundances to "otu" variable
+    otu <- assays(x)[[abund_values]]
+
+    #if index does not have any values
+    if (is.null(index)) {
```

Comment:
The default value of index in .dominance_help (I haven't yet changed the name) is now "all". But there is no "all" index, so it does not really do anything. I changed the default value of index to NULL, because I think it makes more sense, because then the index is calculated based on other values that are provided by user (or from default).

Reply:
change the default to all the indeces and remove the `all` value. Then all the indices will be calculated

Reply:
Like this `index = c("DBP", "DMN", "absolute", "relative", "simpson", "core_abundance", "gini")`. add this in the generic or add all the values in `setMethod` and the `match.arg` call

Reply:
Sorry, if I did not understood correctly, but now the .dominance_help is just calculating one index at a time, so it is not possible to add all indices to .dominance_help's default argument.  It is like .run_dvrsty in estimateDiversity.

Reply:
All in all, estimateDominance works like estimateDiversity. All indices are as a default value of "index" in estimateDominance.

Like this 
----------------------------------------------
#' @rdname estimateDominance
#' @export
setMethod("estimateDominance", signature = c(x = "MicrobiomeExperiment"),
          function(x, abund_values = "counts", index = c("DBP", "DMN", "absolute", "relative", "simpson", "core_abundance", "gini"),
                   ntaxa=1, as_relative=TRUE, aggregate=TRUE, name = index, ..., BPPARAM = SerialParam()){
----------------------------------------------

Reply:
And as a default, you can calculate all indices, like with estimtaeDiversity.

#with...
esophagus <- estimateDominance(esophagus)
esophagus <- estimateDiversity(esophagus)
#...you can calculate all dominance and diversity indices which can be accessed with 
colData(esophagus)

Reply:
I implemented major part of estimateDiversity to estimateDominance, so only bigger difference right now is, how .dominance_help/.run_dvrsty work in detail even though they work same way from the point of user. 

E.g. there is now...

#Calculates dominance indices
              dominances <- BiocParallel::bplapply(X=index,
                            FUN = .dominance_help, x=x,
                            abund_values=abund_values, ntaxa=ntaxa,
                            as_relative=as_relative,
                            aggregate=aggregate,
                            BPPARAM = BPPARAM)

              #Add dominance indices to colData
              .add_dominances_values_to_colData(x, dominances, name)


...which is "directly" from estimateDiversity.

Reply:
So, do you think that also .dominance_help should get all indices as default? That would require some kind of loop/apply structure

Reply:
Lets take a step back.

1. Is the .domincance_help function still necessary?
2. If yes what is the function of `.domincance_help`?
3. Why does `.domincance_help` need to know the value of `index`?
4. If it needs to know, than it needs to be able to handle indeces with length of `>= 1`.

So I think somewhere in this line of inquiry lies the solution. It might help, if you would push the current state into the branch, so it gets updated here and I can have a look at the current state.

Otherwise keep up the good work! 👍

Reply:
Oh, I did not realize that it is not updating here. I was updating only my own branch... sorry. Now, all changes should be here also

Reply:
I just think that we should follow the identical conventions with estimateDominance and estimateDiversity wherever possible.

Reply:
I agree. I am quite sorry, that I had to change things around myself, but the dominance values are getting computed in several different ways, so that it was very complex to explain. I hope you both don't mind the changes

Reply:
On my behalf, it is more than fine that you did those changes. Actually, I was thinking if I should implement that "last part" of estimateDiversity also because of the consistency, and there was no reason to do it differently in estimateDominance. So yeah, everything is fine

Reply:
Cool 👍 

btw: the `.add_dvrsty_values_to_colData` function is a bit more complex, because it needs to handle nested `DataFrame`s. Since the dominance calculation always return a vector, that is not necessary